roles/translator/src/lib/mod.rs

```diff
@@ -1,7 +1,280 @@
+                }
+            }
+
+            // Start receiving messages from the SV2 Upstream role
+            if let Err(e) = upstream_sv2::Upstream::parse_incoming(upstream.clone()) {
+                error!("failed to create sv2 parser: {}", e);
+                return;
+            }
+
+            debug!("Finished starting upstream listener");
+            // Start task handler to receive submits from the SV1 Downstream role once it connects
+            if let Err(e) = upstream_sv2::Upstream::handle_submit(upstream.clone()) {
+                error!("Failed to create submit handler: {}", e);
+                return;
+            }
+
+            // Receive the extranonce information from the Upstream role to send to the Downstream role
+            // once it connects also used to initialize the bridge
+            let (extended_extranonce, up_id) = rx_sv2_extranonce.recv().await.unwrap();
+            loop {
+                let target: [u8; 32] = target.safe_lock(|t| t.clone()).unwrap().try_into().unwrap();
+                if target != [0; 32] {
+                    break;
+                };
+                async_std::task::sleep(std::time::Duration::from_millis(100)).await;
+            }
+
+            let task_collector_bridge = task_collector_init_task.clone();
+            // Instantiate a new `Bridge` and begins handling incoming messages
+            let b = proxy::Bridge::new(
+                rx_sv1_downstream,
+                tx_sv2_submit_shares_ext,
+                rx_sv2_set_new_prev_hash,
+                rx_sv2_new_ext_mining_job,
+                tx_sv1_notify.clone(),
+                status::Sender::Bridge(tx_status.clone()),
+                extended_extranonce,
+                target,
+                up_id,
+                task_collector_bridge,
+            );
+            proxy::Bridge::start(b.clone());
+
+            // Format `Downstream` connection address
+            let downstream_addr = SocketAddr::new(
+                IpAddr::from_str(&proxy_config.downstream_address).unwrap(),
+                proxy_config.downstream_port,
+            );
+
+            let task_collector_downstream = task_collector_init_task.clone();
+            // Accept connections from one or more SV1 Downstream roles (SV1 Mining Devices)
+            downstream_sv1::Downstream::accept_connections(
+                downstream_addr,
+                tx_sv1_bridge,
+                tx_sv1_notify,
+                status::Sender::DownstreamListener(tx_status.clone()),
+                b,
+                proxy_config.downstream_difficulty_config,
+                diff_config,
+                task_collector_downstream,
+            );
+        }); // End of init task
+        let _ =
+            task_collector.safe_lock(|t| t.push((task.abort_handle(), "init task".to_string())));
+    }
+
+    pub fn upstream_address(&self) -> SocketAddr {
+        self.config.upstream_address.clone().parse().unwrap()
+    }
```

Comment:
Where is this used?

Reply:
here: https://github.com/jbesraa/stratum/blob/42b11af73c19bc1ba4158cccc22fbd668d7548e4/roles/tests-integration/tests/integration_e2e.rs#L35